hoover_script.py

```diff
@@ -30,9 +30,167 @@
     raise RuntimeError("HOOVER_HOME environment variable is not set")
 home = Path(_home)
 
-def question(label, default):
-    rv = input("{} [{}]: ".format(label, default))
-    return rv.strip() or default
+interactive_mode = False
+
+class Param:
+    param_list = []
+
+    def __init__(self, name, default, environ=None, question_label=None, optional=False):
+        self.name = name
+        self.default = default
+        self.environ = environ
+        self.question_label = question_label
+        self.optional = optional
+        self.value = None
+        Param.param_list.append(self)
+
+    @staticmethod
+    def _question(label, default):
+        rv = input("{} [{}]: ".format(label, default))
+        return rv.strip() or default
+
+    def get(self):
+        if self.value is not None:
+            return self.value
+
+        if os.getenv(self.environ):
+            self.value = os.getenv(self.environ)
+        elif interactive_mode and question_label is not None:
+            self.value = _question(self.question_label, self.default)
+        else:
+            self.value = self.default
+
+        if self.value is None and not self.optional:
+            raise RuntimeError(("The {} param was not set! Use the {} " +
+                "environment variable to set it.").format(self.name, self.environ))
+
+        return self.value
+
+    def get_path(self):
+        return Path(self.get())
+
+VIRTUALENV_URL = Param(
```

Comment:
Uppercase variable names are reserved for constants. This is more like a singleton. If you want, you can group them in a class, to create an ad-hoc namespace:

```python
class Params:
    virtualenv_url = Param(...)
```